Taking Flight to the Land of Off-Heap, Act I: Page Meta-Data

chrisvest opened the pull request:
* The `MuninnPage` objects are gone.
* In their place is a new `PageList`.
* The `PageList` allocates and initialises a list of pages up, with IDs 0 to N.
* To work with a page, it must first be dereferenced, turning a _page ID_ (`int`) into a _page reference_ (`long`).
* All on-heap page data is immutable; all mutable page meta-data state is now off-heap.
* This reduces the memory usage for page meta-data from 64 bytes (including object headers) to 32 bytes.
* `MuninnPagedFile` now extends `PageList` via a copy-constructor, removing the `pageCache` indirection between it, and the set of memory pages.
* The `SequenceLock` is gone, replaced by an `OffHeapSequenceLock` which works on pointers to 8 byte lock-words.
* The new sequence lock also takes over the responsibility of managing the "dirty bit", or modified flag, as it is now called. This removes a memory write in maintaining page dirtiness.
* We can no longer store _references_ to page swappers in the pages, so we now indirect with an ID into a `SwapperSet`.
* The `SwapperSet` is now in charge of the life cycle of the page swappers, and their IDs, which can be reused.
* Of note, a swapper ID cannot be reused until all pages that uses it, have been evicted. This will only happen after the swapper has been closed.
* _Currently there is no vacuuming of swapper IDs, this needs to be added._
* The `MuninnPagedFile` translation tables are now `int[][]` arrays instead of `Object[][]`, to refer to _page IDs_.
* Since the translation table can now no longer contain references to `BinaryLatch` instances during page faulting the given entry, a new `LatchMap` is introduced to hold and manage latches for page faulting.
* The memory used for the page meta-data in the `PageList` is currently _NOT_ accounted against the memory allocated to the page cache. It is worth considering, though, as I think it simplifies capacity planning wrt. to memory consumption.

chrisvest committed:
Optimise MemoryManager for huge allocations

The MemoryManager would previously throw away the entire current slab if it was asked to perform an allocation that it cannot service.
Now, if the allocation is greater than the grab size, the MemoryManager will perform the allocation in a dedicated slab for the huge allocation, without throwing away the current slab.
This way, the memory in the current slab is not wasted.

chrisvest committed:
First draft of off-heap PageList implementation

chrisvest committed:
Make the PageSwapper use longs and ints instead of Page objects

for reading and writing data to/from pages in memory, to/from sections in
files.

chrisvest committed:
List of TODOs for implementing eviction in PageList

chrisvest committed:
More work on eviction via PageList

chrisvest committed:
Add SwapperSet to manage PageSwappers for PageList

chrisvest committed:
Expand the SwapperSet API with a couple of useful additions

chrisvest committed:
Simplify the SwapperSet interface

chrisvest committed:
Update PageList to make use of the new SwapperSet, and starton implementing eviction

chrisvest committed:
Complete the first draft implementation of PageList.tryEvict

chrisvest committed:
Remove the TODO for PageList.flush because it is not at present obvious if it is needed

chrisvest committed:
Switch to off-heap PageList

The PageCacheTests pass, but the PageCacheSlowIT do not. Yet.

chrisvest committed:
Fix mistake in StubPageCursor porting to ByteBuffer

chrisvest committed:
Fix bug in PageList.evict

Pages that are loaded but not bound, do not have a swapper associated with them.
Looking for a swapper that is not there throws an exception, so we now avoid doing that.

chrisvest committed:
Fix race bug in page faulting

If a thread observes a missing entry in the translation table, it will attempt to initiate a page fault.
Only one thread at a time should try to fault any given page, and this is guarded by the LatchMap.
However, it can be the case that another thread completes the fault in between us noticing that a fault is needed, and trying to grab a latch.
Since the other thread completed the page fault, and released its latch, we will get our own latch and mistakenly do another page fault.
We now guard for this by double-checking that the entry is indeed still missing after we have gotten our latch.
If the entry is no longer missing, then it means that another thread completed the fault ahead of us, and so we simply release the latch we got and retry our pin.

chrisvest committed:
Page cache test updates that are helpful for debugging

chrisvest committed:
Fix a test that failed on a int/long type mismatch

chrisvest committed:
Fix ReadPageCursor shouldRetry bug with linked cursors

When we checked the lock validity using the stamp from the top-most page cursor, instead of the stamp for the given cursor.
This obviously often resulted in the check failing, with no way to get out of the shouldRetry look.
The shouldRetry check now currectly uses the lock stamp for the given cursor in the chain, and also makes sure to refresh the lock state of all the cursors in the chain when a restart is necessary.

chrisvest committed:
Minor optimisation of UnsafeUtil native access check

chrisvest committed:
Formatting

chrisvest committed:
Speed up bounds checking on non-offsetted page accesses

This is possible because we can ignore the underflow case when using the implied offset

chrisvest committed:
PageCursor.setOffset now bounds checks

This is to uphold the assumptions of the nextBoundedPointer method, which expects the offset field to at least never underflow.

chrisvest committed:
Fix a generics raw-type warning

chrisvest committed:
Use branching for bounds checking

It turns out that branching code is faster than branch free code in this case.
CPUs are really good at branch prediction, I guess.
Measurements show the difference is very slight, but it is none the less there.

chrisvest committed:
Remove dead code

chrisvest committed:
Update page cache code comments

chrisvest committed:
PageList page locks are initialised with exclusive lock held

This is useful because pages that are free in the page cache, have their exclusive locks taken.
So this speeds up the initial allocation of the page cache.

chrisvest committed:
New page lock transition: write->flush

It's now possible to atomically undo a write lock, and attempt to grab a flush
lock, on a page in PageList.
This will be useful for adding a PF flag that instructs write page cursors to
flush the page on unpin.
This in turn will be useful for certain stages or phases in the parallel batch
importer, where a group of reader threads are reading in pages ahead of a group
of writer threads who then modify those pages. Currently the reads are fielding
most of the page faulting, and thus also end up doing most of the flushing.
That is, the reads end up doing both the read and the write IO.

chrisvest committed:
PageCacheTest cleanup

chrisvest committed:
Add PF_EAGER_FLUSH

chrisvest committed:
Fix bug in shouldRetry with closed linked cursor

When we discover, via shouldretry, that a cursor needs to retry, the retry will
cascade to all linked cursors, if any.
If the cursor has previously had a linked cursor, but the linked cursor was
closed, then the parent cursor is not notified and will still traverse through
the closed linked cursor when starting retry.
This would cause the startRetry to throw a NullPointerException, because
a closed cursor drops its reference to the MuninnPagedFile.
The bug is fixed by checking that the pinned page ref is still valid, which is
the standard way of telling whether a cursor has been closed or not.
This issue was discovered by GBPTreeConcurrencyIT, which deterministically
failed.
PageCacheTests have been added in a different PR because I wanted them to go
into 3.2, instead of 3.3.

ragadeeshu commented on community/unsafe/src/main/java/org/neo4j/unsafe/impl/internal/dragons/MemoryManager.java:
```diff
@@ -62,6 +66,22 @@ public MemoryManager( long expectedMaxMemory, long alignment )
+            // This is a huge allocation. Put it in its own slab and keep any existing slab at the head.
+            Slab nextSlab = slabs == null? null : slabs.next;
+            Slab allocationSlab = new Slab( nextSlab, bytes, alignment );
+            if ( !allocationSlab.canAllocate( bytes ) )
```
I think that this needs a comment. The first call works if stuff are perfectly aligned? And if the first call doesn't work you need `[1,alignment]` bytes more, and thus you add `alingment` to be safe?

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java:
```diff
@@ -0,0 +1,354 @@
+ * at the start of the critical section, which is then validated at the end of the critical section. The stamp is
+ * invalidated if any write lock or exclusive lock was overlapping with the read lock.
+ * <p>
+ * The concurrent write locks works by assuming that writes are always non-conflicting, so no validation is required.
```
Why can we assume this?

chrisvest replied:
The entities are protected by the transactional lock manager.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java:
```diff
@@ -0,0 +1,354 @@
+ * static.
+ * <p>
+ * Note that the lock-word is assumed to be 8 bytes, and should ideally be aligned to 8-byte boundaries, and ideally
+ * run on platforms that support 8-byte-wide atomic memory operations.
```
Any examples on platforms that don't support `8-byte-wide atomic memory operations`?

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java:
```diff
@@ -0,0 +1,354 @@
+     */
+    public static boolean validateReadLock( long address, long stamp )
+    {
+        UnsafeUtil.loadFence();
```
Should this not be a storeFence, in order to make sure that the an invalidating store does not get reordered to happen after the folloing load?

chrisvest replied:
You should not use a read lock to coordinate stores.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java:
```diff
@@ -0,0 +1,354 @@
+     */
+    public static boolean tryWriteLock( long address )
+    {
+        long s, n;
```
Just because the code resembles C doesn't mean that you have to adopt horrendous naming standards ;)

chrisvest replied:
Aww… Doug Lea does it in `StampedLock`. Why can't I? :)

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java:
```diff
@@ -0,0 +1,354 @@
+            n = s + CNT_UNIT | MOD_MASK;
+            if ( compareAndSetState( address, s, n ) )
+            {
+                UnsafeUtil.storeFence();
```
And I suppose that this is in order to make sure that everyone sees that we took the lock, yes?

chrisvest replied:
No, this is to prevent plain stores within the critical section from reordering _into_ the compareAndSetState, which is theoretically possible, e.g. if it is implemented with LL/SC, or with a mutex. The plain store could then reorder with the store that is the taking of the lock, thus effectively escaping the critical section.

chrisvest replied:
See http://gee.cs.oswego.edu/cgi-bin/viewcvs.cgi/jsr166/src/main/java/util/concurrent/locks/StampedLock.java?r1=1.62&r2=1.63
And http://jsr166-concurrency.10961.n7.nabble.com/Does-StampedLock-need-a-releaseFence-in-theory-td13563.html

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java:
```diff
@@ -0,0 +1,354 @@
+            n = nextSeq( s ) - CNT_UNIT;
+        }
+        while ( !compareAndSetState( address, s, n ) );
+    }
```
No fence since it will be visible sooner or later, and there is no harm in an unlocked page looking like it is locked?

chrisvest replied:
There is in principle no harm in coarsening a critical section. It's a _roach motel_; you can move in, but you can never move out.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -0,0 +1,332 @@
+            UnsafeUtil.putLong( address += 8, PageCursor.UNBOUND_PAGE_ID ); // file page id
+            UnsafeUtil.putLong( address += 8, 0 ); // rest
+        }
+        UnsafeUtil.fullFence(); // Guarantee the visibility of the cleared memory
```
Would not a `storeFence` be enough here? Since the method only does store operations?

chrisvest replied:
Wearing belt and suspenders here because I'm using plain writes, and this is happening before the `OffHeapPageLock` takes over the responsibility of protecting the pages and their meta-data.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -0,0 +1,332 @@
+        return baseAddress + pageId * META_DATA_BYTES_PER_PAGE;
+    }
+
+    private long offLock( long pageRef )
```
Curious method names

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -0,0 +1,332 @@
+
+    private void setSwapperId( long pageRef, int swapperId )
+    {
+        UnsafeUtil.putInt( offSwapperId( pageRef ), swapperId );
```
No fence for this guy? (And his other `set...` buddies)

chrisvest replied:
No, this state is guarded by the `OffHeapPageLock` in various forms.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -0,0 +1,332 @@
+        // translation table will re-do the page fault.
+        setFilePageId( pageRef, filePageId ); // Page now considered isLoaded()
+        PageObj page = new PageObj( cachePageSize, address( pageRef ) ); // todo avoid this allocation
+        long bytesRead = swapper.read( filePageId, page );
```
The purpose of this line is not immediately clear to me. Care to elaborate?

chrisvest replied:
`swapper.read` is where we read the page from the file into memory.

ragadeeshu commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/PageListTest.java:
```diff
@@ -0,0 +1,1157 @@
+        assertThat( pageList.getFilePageId( pageRef ), is( PageCursor.UNBOUND_PAGE_ID ) );
+    }
+
+    // xxx ---[ Page fault tests ]---
```
For the interference tests, perhaps you should validate all fields of all other entries in the list? Or at least all fields of adjacent entries?

ragadeeshu commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/PageListTest.java:
```diff
@@ -0,0 +1,1157 @@
+        nextPageRef = pageList.deref( nextPageId );
+    }
+
+    // xxx ---[ Sequence lock tests ]---
```
Some concurrent read-invalidation and contention for locks of same & different types would be interesting. Maybe a bit annoying to implement though.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -71,6 +70,22 @@
+     * object.
+     * @param pageList The {@code PageList} instance whose state to copy.
+     */
+    PageList( PageList pageList )
```
Why do we need this, instead of passing the same object? I'm sure that the explanation is in the comment just above this, but I don't quite get it.

chrisvest replied:
It's so `MuninnPagedFile` can extend `PageList` itself, and get all the relevant state carried over from the page list instance created in `MuninnPageCache`.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/SwapperSet.java:
```diff
@@ -0,0 +1,150 @@
+
+final class SwapperSet
+{
+    private volatile Allocation[] allocations = new Allocation[0];
```
You have three different locks in this class. Does this need to be volatile, or could you move it into one of the locks? And is this class such a hotspot that there is need to have three separate locks?

ragadeeshu reviewed (commented):
This is a first batch in order to not loose these comments as I move forward.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/SwapperSet.java:
```diff
@@ -0,0 +1,150 @@
+import org.neo4j.collection.primitive.PrimitiveIntSet;
+import org.neo4j.io.pagecache.PageSwapper;
+
+final class SwapperSet
```
A general comment about the purpose of this class would be nice :)

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/LatchMap.java:
```diff
@@ -0,0 +1,104 @@
+import org.neo4j.unsafe.impl.internal.dragons.FeatureToggles;
+import org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil;
+
+final class LatchMap
```
This class could do with some explanation, I think, both general and method-specific.

ragadeeshu commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -62,6 +62,20 @@
+    // xxx how many times the file page id loop around the translation table before arriving at the given entry.
+    // xxx That is, what the entry index should be multiplied by to get the file page id. To do this, we would, however,
+    // xxx have to either grab bits from the page id space, or make each entry more than 4 bytes. This will all depend
+    // xxx on where the cut-off point is. That is, what the max entry capacity for a translation table should be.
```
Intriguing

chrisvest committed:
Update speculation comment in PageList

chrisvest committed:
Finish SwapperSet and reuse of swapper ids

chrisvest committed:
Fix checkstyle and compilation errors

chrisvest committed:
Javadoc updates for SwapperSet

chrisvest committed:
Fix page cursor usage bug in GBPTree

It was possible for an inconsistent read of a read page cursor to cause an out
of bounds access to a write page cursor, when writing the page header during
state forcing.
The fix is to explicitly clear the bounds flag on the write cursor, for every
shouldRetry iteration of the read cursor.

chrisvest committed:
Javadoc for LatchMap

chrisvest committed:
Fix PageCacheMetricsIT - the page cache is allowed to evict when it pleases

chrisvest committed:
Checkstyle fixes

chrisvest committed:
Don't omit stack traces from exceptions when running tests

chrisvest committed:
Fix TestCrashWithRebuildSlow flakiness caused by background eviction

chrisvest committed:
Fix flaky TransactionBoundQueryContextTest

If there is an eviction between the two getNodeById calls, one of them will be
counted as a miss.
There is apparently also a chance of getting more than two accesses.

chrisvest committed:
Fix MetricsKernelExtensionFactoryIT test flakiness and test bugs

chrisvest committed:
Don't let vacuum cause evictions unless there is a pressing need

Previously, we would try to vacuum as soon as 20 files had been unmapped. Such
activity is common around the startup of the database, when the page cache has
plenty of free pages. This caused unexpected evictions to occur quite early in
the life of a database, which is technically fine, but caused many tests to
fail.

Now we don't do that; if the eviction thread has not already started, then we
won't vacuum until there is a "pressing need" in the sense that there are less
than 200 swapper ids left.

chrisvest committed:
Don't forget the free page counter when switching to free page list

chrisvest committed:
Rename the underlying unit of allocation in the MemoryManager from Slab to Grab, to avoid confusion things with the Slab allocator

chrisvest committed:
Fix compilation after rebase

chrisvest committed:
Fix compilation after cursor revert
